eclair-core/src/main/scala/fr/acinq/eclair/wire/protocol/LiquidityAds.scala

```diff
@@ -0,0 +1,239 @@
+            Left(InvalidLiquidityAdsAmount(channelId, remoteFundingAmount, fundingAmount))
+          } else if (maxFee < fees) {
+            Left(InvalidLiquidityRates(channelId))
+          } else {
+            val leaseAmount = fundingAmount.min(remoteFundingAmount)
+            Right(Lease(leaseAmount, fees, willFund.sig, witness))
+          }
+        case None =>
+          // If the remote peer doesn't want to provide inbound liquidity, we immediately fail the attempt.
+          // The node operator may retry this funding attempt without requesting inbound liquidity.
+          Left(MissingLiquidityAds(channelId))
+      }
+    }
+  }
+
+  def validateLease_opt(requestRemoteFunding_opt: Option[RequestRemoteFunding],
+                        remoteNodeId: PublicKey,
+                        channelId: ByteVector32,
+                        fundingScript: ByteVector,
+                        remoteFundingAmount: Satoshi,
+                        fundingFeerate: FeeratePerKw,
+                        willFund_opt: Option[ChannelTlv.WillFund]): Either[ChannelException, Option[Lease]] = {
+    requestRemoteFunding_opt match {
+      case Some(requestRemoteFunding) => requestRemoteFunding.validateLease(remoteNodeId, channelId, fundingScript, remoteFundingAmount, fundingFeerate, willFund_opt) match {
+        case Left(t) => Left(t)
+        case Right(lease) => Right(Some(lease))
+      }
+      case None => Right(None)
+    }
+  }
+
+  /** We propose adding funds to a channel for an optional fee at the given rates. */
+  case class AddFunding(fundingAmount: Satoshi, leaseRate_opt: Option[LeaseRate]) {
+    def signLease(nodeKey: PrivateKey, fundingScript: ByteVector, fundingFeerate: FeeratePerKw, requestFunds_opt: Option[ChannelTlv.RequestFunds]): Option[WillFundLease] = for {
+      rate <- leaseRate_opt
+      request <- requestFunds_opt
+    } yield {
+      rate.signLease(nodeKey, fundingScript, fundingFeerate, request, Some(fundingAmount))
+    }
+  }
+
+  /**
+   * Liquidity is leased using the following rates:
+   *
+   *  - the buyer pays [[leaseFeeBase]] regardless of the amount contributed by the seller
+   *  - the buyer pays [[leaseFeeProportional]] (expressed in basis points) of the amount contributed by the seller
+   *  - the seller will have to add inputs/outputs to the transaction and pay on-chain fees for them, but the buyer
+   *    refunds on-chain fees for [[fundingWeight]] vbytes
+   *
+   * The seller promises that their relay fees towards the buyer will never exceed [[maxRelayFeeBase]] and [[maxRelayFeeProportional]].
+   * This cannot be enforced, but if the buyer notices the seller cheating, they should blacklist them and can prove
+   * that they misbehaved using the seller's signature of the [[LeaseWitness]].
+   */
+  case class LeaseRate(leaseDuration: Int, fundingWeight: Int, leaseFeeProportional: Int, leaseFeeBase: Satoshi, maxRelayFeeProportional: Int, maxRelayFeeBase: MilliSatoshi) {
```

Comment:
We have to follow the spec encoding order here (because `scodec`), which encodes the proportional fees first. I believe that one of the reasons Lisa chose that order is because we may deprecate the base fee at some point?